src/app/InvokeInteraction.h

```diff
@@ -0,0 +1,438 @@
+        kStateAwaitingResponse = 1,
+    };
+
+    
+    CHIP_ERROR StartCommandHeader(const CommandParams &aParams);
+    void CloseExchange();
+
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * apExchangeContext, const PacketHeader & aPacketHeader,
+                           const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload) override;
+    void OnResponseTimeout(Messaging::ExchangeContext * apExchangeContext) override;
+
+    Messaging::ExchangeContext *GetExchangeContext() { return mpExchangeCtx; }
+    friend class InteractionModelEngine;
+
+    CHIP_ERROR FinalizeMessage(System::PacketBufferHandle &aBuf);
+    CHIP_ERROR SendMessage(System::PacketBufferHandle aBuf);
+  
+private: 
+    enum Mode
+    {
+        kModeUnset = 0,
+        kModeClientInitiator = 1,
+        kModeServerResponder = 2
+    };
+
+    friend class TestInvokeInteraction;
+    bool mExpectsResponses = false;
+    State mState = kStateReady;
+    chip::System::PacketBufferTLVWriter mWriter;
+    app::InvokeCommand::Builder mInvokeCommandBuilder;
+    ICommandHandler *mHandler;
+    Messaging::ExchangeManager *mpExchangeMgr = nullptr;
+    Messaging::ExchangeContext *mpExchangeCtx = nullptr;
+};
+
+/*
+ * @brief
+ *
+ * This class represents an invoke interaction from the responding side (i.e on the server). An instance of this class maps to a specific
+ * invoke interaction over the wire. Consequently, the lifetime of this object is limited to a SINGLE interaction backed by a SINGLE exchange context.
+ *
+ * This object permits a single request + response. It does not permit multiple responses to be sequenced in order.
+ *
+ * This object comes into existence at the point that an invoke request has to be handled. Upon initialization of the object with the packet buffer, this
+ * object parses the various command data elements embedded in the message and dispatches them to ClusterServer objects registered with the 
+ * InteractionModel engine.
+ *
+ * Upon dispatch, the ClusterServer instance's OnInvokeRequest method is invoked. The cluster's OnInvokeRequest implementation has to follow one of the following
+ * calling conventions:
+ *      1. Handle the request and if a response needs to be generated, call AddResponse on the passed in InvokeResponder object.
+ *      2. Handle the request and if no responses needs to to be generated, do nothing.
+ *      3. Handle the request if a response needs to be generated asynchronously, call IncrementHoldOffRef() on the passed in InvokeResponder object. This
+ *         is a signal to the responder object to hold-off sending any queued up responoses till a matching DecrementHoldOffRef() call is made by that handler.
+ *
+ */
+class InvokeResponder
+{
+public:
+    /*
+     * @brief
+     *
+     * This interface defines the expectations for a handler of command requests.
+     *
+     */
+    class CommandHandler
```

Comment:
Sync Notes: This will be removed since the actual dispatch and handling of a command will retain the existing `DispatchSingleClusterResponseCommand` logic.